core/src/storage/impls/state.rs

```diff
@@ -276,9 +285,124 @@ impl<'a> StateTrait for State<'a> {
+            None => None,
+            Some(root_node) => {
+                if self.maybe_intermediate_trie_key_padding.is_some()
+                    && self.maybe_intermediate_trie.is_some()
+                {
+                    let intermediate_trie_key_padding = self
+                        .maybe_intermediate_trie_key_padding
+                        .as_ref()
+                        .unwrap();
+                    let intermediate_mpt_key_prefix = access_key_prefix
+                        .to_delta_mpt_key_bytes(intermediate_trie_key_padding);
+                    let values = SubTrieVisitor::new(
+                        self.maybe_intermediate_trie.as_ref().unwrap(),
+                        root_node.clone(),
+                        &mut self.owned_node_set,
+                    )?
+                    .traversal(
+                        &intermediate_mpt_key_prefix,
+                        &intermediate_mpt_key_prefix,
+                    )?;
+
+                    values
+                } else {
+                    None
+                }
+            }
+        };
+
+        // Retrieve key/value pairs from snapshot
+        let mut kv_iterator = self.snapshot_db.snapshot_kv_iterator();
+        let lower_bound_incl = access_key_prefix.to_key_bytes();
+        let mut upper_bound_excl_value = lower_bound_incl.clone();
+        let upper_bound_excl = if lower_bound_incl.len() == 0 {
+            None
+        } else {
+            let mut carry = 1;
+            let len = upper_bound_excl_value.len();
+            for i in 0..len {
+                if upper_bound_excl_value[len - 1 - i] == 255 {
+                    upper_bound_excl_value[len - 1 - i] = 0;
+                } else {
+                    upper_bound_excl_value[len - 1 - i] += 1;
+                    carry = 0;
+                    break;
+                }
+            }
+            // all bytes in lower_bound_incl are 255, which means no upper bound
+            // is needed.
+            if carry == 1 {
+                None
+            } else {
+                Some(upper_bound_excl_value.as_slice())
             }
+        };
+        let mut kvs = kv_iterator
+            .iter_range(lower_bound_incl.as_slice(), upper_bound_excl)?;
+
+        let mut snapshot_kvs = Vec::new();
+        while let Some((key, value)) = kvs.next()? {
+            snapshot_kvs.push((key, value.into_boxed_slice()));
         }
+
+        let mut result = Vec::new();
+        // This is used to keep track of the deleted keys.
+        let mut deleted_keys = HashSet::new();
+        if let Some(kvs) = delta_trie_kvs {
+            for (k, v) in kvs {
+                deleted_keys.insert(k.clone());
+                if v.len() > 0 {
+                    result.push((k, v));
+                }
+            }
+        }
+
+        if let Some(kvs) = intermediate_trie_kvs {
+            for (k, v) in kvs {
+                let storage_key = StorageKey::from_key_bytes(&k);
```

Comment:
This key is a "delta_mpt_key". from_key_bytes only takes the storage_key, which is === the actual key.